zk/stages/stage_sequence_execute_data_stream.go

```diff
@@ -46,31 +48,54 @@ func (sbc *SequencerBatchStreamWriter) CommitNewUpdates() ([]*verifier.VerifierB
 
 func (sbc *SequencerBatchStreamWriter) writeBlockDetailsToDatastream(verifiedBundles []*verifier.VerifierBundle) ([]*verifier.VerifierBundle, error) {
 	var checkedVerifierBundles []*verifier.VerifierBundle = make([]*verifier.VerifierBundle, 0, len(verifiedBundles))
+
 	for _, bundle := range verifiedBundles {
 		request := bundle.Request
 		response := bundle.Response
 
 		if response.Valid {
-			parentBlock, err := rawdb.ReadBlockByNumber(sbc.sdb.tx, request.GetLastBlockNumber()-1)
+			highestClosedBatch, err := sbc.streamServer.GetHighestClosedBatch()
+			if err != nil {
+				return checkedVerifierBundles, err
+			}
+			highestStartedBatch, err := sbc.streamServer.GetHighestBatchNumber()
+			if err != nil {
+				return checkedVerifierBundles, err
+			}
+			isCurrentBatchHigherThanLastInDatastream := request.BatchNumber > highestStartedBatch
+			isLastBatchInDatastremClosed := highestClosedBatch == highestStartedBatch
+			if isCurrentBatchHigherThanLastInDatastream && !isLastBatchInDatastremClosed {
+				if err := finalizeLastBatchInDatastream(sbc.batchContext, highestStartedBatch, request.GetLastBlockNumber()-1); err != nil {
```

Comment:
Do we not want to close it at the very first block number in the request -1 rather than the last?

Reply:
absolutely correct